File: baekjoon/DFS&BFS/1388_바닥 장식.py
N, M = map(int, input().split(" "))
graph = [list(input()) for _ in range(N)]


def division(i, j, cnt):
    compare = graph[i][j]
    if compare == '-':
        for k in range(j, M):
            if graph[i][k] == compare:
                graph[i][k] = cnt
            else:
                return
    elif compare == '|':
        for k in range(i, N):
            if graph[k][j] == compare:
                graph[k][j] = cnt
            else:
                return


def solution():
    cnt = 0
    for i in range(N):
        for j in range(M):
            if type(graph[i][j]) == str:
                division(i, j, cnt)
                cnt += 1
    return cnt


print(solution())

File: baekjoon/DFS&BFS/1388_바닥 장식.js
const fs = require('fs')
const [nm, ...input] = fs
  .readFileSync('../input.txt')
  .toString()
  .trim()
  .split('\n')
const [N, M] = nm.split(' ').map(Number)
const graph = input.map((x) => x.trim().split(''))

const division = (i, j, cnt) => {
  const compare = graph[i][j]

  if (compare === '-') {
    for (let k = j; k < M; k++) {
      if (graph[i][k] === compare) {
        graph[i][k] = cnt
      } else {
        break
      }
    }
  } else if (compare === '|') {
    for (let k = i; k < N; k++) {
      if (graph[k][j] === compare) {
        graph[k][j] = cnt
      } else {
        break
      }
    }
  }
}

function solution() {
  let cnt = 0
  graph.forEach((row, i) => {
    row.forEach((column, j) => {
      if (isNaN(column)) {
        division(i, j, cnt)
        cnt += 1
      }
    })
  })

  return cnt
}

console.log(solution())
